tests/conftest.py

```diff
@@ -5,6 +5,9 @@
 
 from guidance import models
 
+# Ensure that asserts from tests/utils.py are rewritten by pytest to show helpful messages
```

Comment:
I've not come across this before, and the `register_assert_rewrite()` docs are not immediately enlightening. Can you describe what this does?

Reply:
Yes, stumbled across it for the first time yesterday actually.

My understanding is that pytest reaches into your testing code (anything matching `test*.py`) and makes some changes before actually running your tests. It may do other things too, but the most obvious and useful one is that it rewrites your asserts to ensure you get helpful error messages.

Say in `test.py` we have the following code:
```python
def number():
    return 3

def check_example():
    assert number() == 4

def test_example():
    check_example()
```

Pytest is very friendly to us here and actually tells us the "contents" of the failed assert.
![image](https://github.com/guidance-ai/guidance/assets/927410/014c8675-ba7c-4e5f-aa62-5d90815f93ac)

Now, if `check_example` were imported from somewhere else (specifically, not from a `test*.py` file), we get a much less helpful error message. Pytest didn't know about the file you imported from when it was collecting tests, so it doesn't do the helpful rewrite of the asserts in the imported code.
![image](https://github.com/guidance-ai/guidance/assets/927410/877b7749-02ba-45f8-a4a9-82f6e641d128)

`register_assert_rewrite` will let you register "external" files and make sure that pytest rewrites the asserts there too. Note that it needs to be called before the file is imported (I think it can be done in the tests `__init__` if we don't want it in conftest).

Reply:
A few minutes after posting, it occurred to me that it might be something like that. I knew that sometimes assert messages were more helpful than others, but I'd not dug into it enough to find the above.